packages/core/src/workflow/piece-add.js

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * @typedef {import('@web3-storage/data-segment').PieceLink} PieceLink
+ * @typedef {import('../data/types.js').Piece<PieceLink>} Piece
+ * 
+ * @typedef {object} EncodedRecord
+ * @property {string} body
+ * @property {string} id
+ * 
+ * @typedef {object} Record
+ * @property {Piece} piece
+ * @property {string} id
+ */
+
+/**
+ * Invokes aggregate/add on the queued system to go through the backlog of pieces provided by Storefront actors.
+ * Once the `aggregate/add` succeeds, the peice is queued for buffering on its way to be aggregated.
+ * When only a subset of the pieces succeed, a `batchItemFailures` is returned so that these items can be re-queued.
+ *
+ * @param {object} props
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Piece>} props.queueClient
+ * @param {import('@web3-storage/filecoin-client/types').InvocationConfig} props.invocationConfig
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} props.aggregatorServiceConnection
+ * @param {EncodedRecord[]} props.records - message encoded piece records
+ */
+export async function addPieces ({
+  queueClient,
+  invocationConfig,
+  aggregatorServiceConnection,
+  records,
+}) {
+  const decodedRecords = await Promise.all(
+    records.map(async pr => ({
+      piece: await decodePiece.message(pr.body),
+      id: pr.id
+    }))
+  )
+
+  let responses
+  try {
+    responses = await Promise.all(
+      decodedRecords.map(record => addPiece({
+        queueClient,
+        invocationConfig,
+        aggregatorServiceConnection,
+        record
+      }))
+    )
+  } catch {
+    return {
+      error: new AggregateAddFailed('failed to add pieces')
+    }
+  }
+
+  return {
+    ok: {
+      // number of successful handled pieces
+      countSuccess: responses.reduce((acc, value) => {
+        return value.ok ? acc + 1 : acc
+      }, 0),
+      // failed pieces that can be re-queued
+      batchItemFailures: responses.reduce((filtered, value) => {
+        if (value.error) {
+          filtered.push(value.id)
+        }
+
+        return filtered
+      }, /** @type {string[]} */ ([]))
+    }
+  }
+}
+
+/**
+ * @param {object} props
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Piece>} props.queueClient
+ * @param {import('@web3-storage/filecoin-client/types').InvocationConfig} props.invocationConfig
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} props.aggregatorServiceConnection
+ * @param {Record} props.record 
+ */
+async function addPiece ({
+  queueClient,
+  invocationConfig,
+  aggregatorServiceConnection,
+  record
+}) {
+  const aggregateAddResponse = await Aggregator.aggregateAdd(
+    invocationConfig,
+    record.piece.piece,
+    record.piece.storefront,
+    record.piece.group,
+    { connection: aggregatorServiceConnection }
+  )
+  if (aggregateAddResponse.out.error) {
+    return {
+      id: record.id,
+      error: aggregateAddResponse.out.error
+    }
+  }
```

Comment:
Please don't overload `Result` type with more fields, TS inference will fail somewhere if you do. More idiomatic approach is to wrap like

```suggestion
  if (aggregateAddResponse.out.error) {
    return {
      error: {
        id: record.id,
        cause: aggregateAddResponse.out.error
      }
    }
  }
```

Reply:
Alternatively you can just return `aggregateAddResponse.out` and let the caller deal with wrapping
`addPiece(record).then(r => [record.id, r])` or something along those lines